packages/next/src/client/app-dir/link.tsx

```diff
@@ -204,6 +204,24 @@ function formatStringOrUrl(urlObjOrString: UrlObject | string): string {
   return formatUrl(urlObjOrString)
 }
 
+/**
+ * Returns the ref of a React element handling differences between React 19 and older versions.
+ * It will throw runtime error if the element is not a valid React element.
+ * @param element React.ReactElement
+ * @returns React.Ref<any> | undefined
+ */
+function getReactElementRef(
+  element: React.ReactElement
+): React.Ref<any> | undefined {
+  // 'ref' is passed as prop in React 19, whereas 'ref' is directly attached to children in older versions
+  if (parseInt(React.version, 10) >= 19) {
```

Comment:
I wasn't sure how you guys want to check this, another option is
```suggestion
  if (element.props.propertyIsEnumerable('ref')) {
```

Reply:
`Object.prototype.propertyIsEnumerable.call(element.props, 'ref')` is preferable.

Reply:
Actually, let's just use the one Radix-UI uses. That one is tested already: https://github.com/radix-ui/primitives/blob/dae8ef4920b45f736e2574abf23676efab103645/packages/react/presence/src/Presence.tsx#L173-L190

Reply:
No objections. Maybe not ideal for bundle size though? https://github.com/radix-ui/primitives/pull/2934#discussion_r1664438772.

Material-UI used: https://github.com/mui/material-ui/pull/43132/files#diff-2aabece439eee5bb9309cc46b1b3c3b690ee8a7231cbac1ba495cf832e506152R16 until it changed to copy Base UI to reduce bundle size.
Base UI uses: https://github.com/mui/base-ui/blob/a14ea0e7600e99a9c052978ee916e3348638b796/packages/react/src/utils/getReactElementRef.ts#L12

Reply:
Testing for React.version is not robust. Feature detection is always preferable. And in that regard the one in Radix UI is actually battle tested. We can optimize later for implementations that have been tested against the full support matrix.

Reply:
Ok, updated. It seems to add +577 B uncompressed and +202 B compressed according to https://github.com/vercel/next.js/pull/74334#issuecomment-2773422163.